crates/autopilot/src/on_settlement_event.rs

```diff
@@ -302,3 +202,230 @@ impl Inner {
+            return Ok(false);
+        };
+        let domain_separator = self.eth.contracts().settlement_domain_separator();
+
+        let (auction_id, auction_data) = match Self::recover_auction_id_from_calldata(
+            &mut ex,
+            &transaction,
+            &model::DomainSeparator(domain_separator.0),
+        )
+        .await?
+        {
+            AuctionIdRecoveryStatus::InvalidCalldata => {
+                // To not get stuck on indexing the same transaction over and over again, we
+                // insert the default auction ID (0)
+                (Default::default(), None)
+            }
+            AuctionIdRecoveryStatus::DoNotAddAuctionData(auction_id) => (auction_id, None),
+            AuctionIdRecoveryStatus::AddAuctionData(auction_id, settlement) => (
+                auction_id,
+                Some(
+                    self.fetch_auction_data(settlement, auction_id, &transaction, &mut ex)
+                        .await?,
+                ),
+            ),
+        };
+
+        let update = SettlementUpdate {
+            block_number: event.block_number,
+            log_index: event.log_index,
+            auction_id,
+            auction_data,
+        };
+
+        tracing::debug!(?hash, ?update, "updating settlement details for tx");
+
+        Postgres::update_settlement_details(&mut ex, update.clone())
+            .await
+            .with_context(|| format!("insert_settlement_details: {update:?}"))?;
+        ex.commit().await?;
+        Ok(true)
+    }
+}
+
+#[allow(dead_code)]
+pub struct CircuitBreaker {
+    circuit_breaker: infra::blockchain::circuit_breaker::CircuitBreaker,
+    persistence: infra::Persistence,
+    db: Postgres,
+    solvers: Vec<eth::Address>,
+    // Registry to keep track of the already processed settlements, format (block, hash)
+    registry: Arc<Mutex<BTreeMap<i64, H256>>>,
+}
+
+impl CircuitBreaker {
+    /// Maximum capacity of the registry in order not to overflow the memory
+    const MAX_REGISTRY_SIZE: usize = 100;
+    /// The number of settlements taken from the database (ordered by block)
+    const NUMBER_SETTLEMENTS: i64 = 4;
+
+    pub fn build(
+        eth: infra::blockchain::circuit_breaker::CircuitBreaker,
+        persistence: infra::Persistence,
+        solvers: Vec<eth::Address>,
+        db: Postgres,
+    ) -> Arc<dyn OnSettlementEventAction> {
+        Arc::new(Self {
+            circuit_breaker: eth,
+            persistence,
+            db,
+            solvers,
+            registry: Arc::new(Mutex::new(BTreeMap::new())),
+        })
+    }
+}
+
+#[async_trait]
+impl OnSettlementEventAction for CircuitBreaker {
+    async fn apply(&self) -> Result<bool> {
+        let mut ex = self
+            .db
+            .pool
+            .begin()
+            .await
+            .context("acquire DB connection")?;
+        let mut events = database::settlements::get_settlements_without_auction(
```

Comment:
We should not use `get_settlements_without_auction` here. It could be that all legitimate auctions have been processed and there are no auctions left but we still got a new settlement. If this settlement was executed out of competition the associated solver needs to get disabled as well.